From 1f78da851b7286a3072ff1a683553f58e0901617 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 09:51:40 +0200
Subject: [PATCH 452/553] Suppress Clang 21 -Wcharacter-conversion
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The new Clang 21 warning is wreaking havoc in C++ projects that want
to use modern C++, because it not only warns about char8_t → char16_t,
but also about the completely harmless char16_t → char32_t conversions.

It has since been fixed, but suppress the warning globally until we
know which version has the fix (looks to be 21.1.4 at this point).

Upstream bug report: https://github.com/llvm/llvm-project/issues/163719
More discussion in: https://github.com/llvm/llvm-project/issues/138526
Upstream fix: https://github.com/llvm/llvm-project/pull/163927

Pick-to: 6.8 6.5
Change-Id: I6409f6f6833131e3a77d81ab679125b2806e86a2
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit f20f12baab1f40ec0aad2635186e7d270139509b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/global/qcompilerdetection.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/corelib/global/qcompilerdetection.h b/src/corelib/global/qcompilerdetection.h
index 700c59ab3c7..df55baf3120 100644
--- a/src/corelib/global/qcompilerdetection.h
+++ b/src/corelib/global/qcompilerdetection.h
@@ -1443,6 +1443,10 @@ QT_WARNING_DISABLE_MSVC(4706) /* assignment within conditional expression */
 QT_WARNING_DISABLE_MSVC(4355) /* 'this' : used in base member initializer list */
 QT_WARNING_DISABLE_MSVC(4710) /* function not inlined */
 QT_WARNING_DISABLE_MSVC(4530) /* C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc */
+#  elif defined(Q_CC_CLANG_ONLY)
+#    if Q_CC_CLANG >= 2100
+        QT_WARNING_DISABLE_CLANG("-Wcharacter-conversion") /* until https://github.com/llvm/llvm-project/issues/163719 is fixed */
+#    endif
 #  elif defined(Q_CC_BOR)
 #    pragma option -w-inl
 #    pragma option -w-aus
-- 
2.51.2

